[#4] apply thread pool

toyseed opened the pull request:
rust programming book 내용을 바탕으로 thread pool 을 적용했습니다. 
코드를 거의 그대로 가져왔어요.

toyseed committed:
[#4] apply thread pool

it is based on the "Rust Programming Book"

toyseed committed:
Merge remote-tracking branch 'upstream/master' into seo/multi-thread

# Conflicts:
#	src/server/server.rs
#	src/server/server_builder.rs

NamsooCho commented on src/http/http_request_header.rs:
```diff
@@ -0,0 +1,18 @@
+    fn from(req: HttpRequestHeader) -> Self {
+        String::clone(&req._raw)
+    }
+}
```
빈행을 맨끝에 한라인 추가해 주세요.

toyseed replied:
빈행 추가했습니다.

NamsooCho commented on src/worker/worker.rs:
```diff
@@ -0,0 +1,52 @@
+    thread::spawn(move || {
+        loop {
+            let mutex = match receiver.lock() {
+                Ok(mutex) => mutex,
```
match의 arm이 몇개 없으면 if let 을 검토해보세요

toyseed replied:
if let 으려 하려니까 if let block 엔에 따로 로직이 없어서 match 유지 했습니다. ^^;

toyseed replied:
if let 으로 수정했습니다.

NamsooCho commented on src/worker/worker_manager.rs:
```diff
@@ -0,0 +1,71 @@
+}
+
+#[cfg(test)]
+mod tests {
```
public 함수들을 테스트하는 코드도 작성해 주시면 좋아요.

toyseed replied:
request 메서드에도 테스트 함수 추가했습니다. task 에서 panic 을 일으켜서 task 가 실행되는지 확인했습니다. 테스트를 작성하다보니까 request 메서드 파라매터로 Box<dyn Task> 대신 Arc<dyn Task> 를 사용하면 테스트가 좀 더 편했을려나 싶기도 하네요.

NamsooCho reviewed (commented):
Great!!!

toyseed committed:
[#4] changes for the code review

- add trailing new line on http_request_header.js
- add test case for the pub fn

toyseed committed:
Merge branch 'seo/init_base_structure' into seo/multi-thread

NamsooCho commented on src/http/http_request_header.rs:
```diff
@@ -0,0 +1,18 @@
+/// model for http request header
+pub struct HttpRequestHeader {
+    _raw: String,
```
_ 접두사를 붙이는건 사용하지 않는 변수라는 뜻입니다.

toyseed replied:
테스트 용도로 사용된 변수라 _ 를 사용했습니다. http 파싱 로직을 작성하면서 삭제할 예정입니다. 차후에는 이런 경우 주석을 달겠습니다.

NamsooCho commented on src/server/server.rs:
```diff
@@ -27,16 +39,20 @@ impl Server {
-            stream.read(&mut buffer)?;
-
-            println!("input: {}", String::from_utf8_lossy(&buffer[..]));
+            let stream = stream.unwrap();
```
실패할지도 모르는 연산에 unwrap을 사용하면 프로세스가 panic합니다.
프로그램은 연산이 실패하면 에러를 리포트하고 적절한 에러처리를 해야 합니다.
절대로 실패하지 않는 연산에는 사용이 가능하며,
test 함수에서는 의도적으로 unwrap을 사용하기도 합니다.

toyseed replied:
unwrap 을 사용하지 않도록 변경했습니다.

NamsooCho reviewed (commented):
많이 나아진 코드입니다.

toyseed committed:
[#4] changes for the code review

- remove unwrap
- change 'match' to 'if let' for simple arms

NamsooCho commented on src/worker/task.rs:
```diff
@@ -0,0 +1,3 @@
+pub trait Task: Send {
+    fn execute(&mut self);
+}
```
맨 마지막에 빈 행을 추가해서 끝나야 합니다.

toyseed replied:
빈 행 추가했습니다. 이참에 소스 코드 마지막에 빈 행이 추가되도록 ide 설정도 변경했습니다.

NamsooCho reviewed (commented):
거의 다 왔어요..

toyseed committed:
[#4] changes for the code review

- add empty line to the end of the file
